docs-internal/db.md

```diff
@@ -0,0 +1,364 @@
+# Helidon Database Proposal
+
+Provide an API that enables reactive database support to be used with Helidon SE.
+
+## Proposal
+
+The Helidon DB is an abstraction layer over
+- database configuration
+- statement configuration
+- statement processing
+- handling of results
+
+The Helidon DB in addition supports
+- queries using either indexed or named parameters
+- mapping of POJOs to database parameters using provided mapper(s)
+    - `.createNamedDmlStatement("insert-mon").namedParam(pokemon)` would work if `Pokemon` db mapper is registered
+- mapping of database query results to POJOs using provided mapper(s)
+    - `dbRow.as(Pokemon.class)` would work if `Pokemon` db mapper is registered
+- mapping of database columns to arbitrary types using provided mappers(s)
+    - `column.as(Long.class)` would work even for String column, as long as `String -> Long` mapper is registered
+   
+
+The Helidon DB is *NOT*
+- Statement abstraction 
+    - users define statements in the language understood by the database (such as `SQL`)
+    - there is *NO* query language other than the database native query language
+    
+As part of this proposal there is also added support for [generic mapping](generic-mapping.md)
```

Comment:
I have mixed feelings about the _generic mapping_ idea. Seems a bit too generic to be useful; perhaps it needs a bit more motivation? Automatic mapping can have some interesting side effects that are hard to find. OTOH, `DbMapper` makes perfect sense.

Reply:
We do have similar APIs in 2 places already (config & webserver) ; I like the idea of having a generic API for converting types, even if that's remains only an internal API.

If a built-in converter for DbRow -> Publisher<DataChunk> is provided for common formats this would nice, (e.g. JSON, XML).

Maybe we accommodate the potential side effects by providing a way to supply different converters when needed.

Reply:
There are a few places this is useful in the DB itself. I think that similar approach could be used by Config.
And yes, the possibility to add additional mappings is available (using a builder).
I agree that this may be interesting when multiple modules use it (e.g. String -> DateTime) mapping - we just have to be consistent in our approach and if we need a String -> DateTime that is unusual, it should be specified in a module specific mapping (as I have done with the DbMapper in DB implementation)

Reply:
Closing this comment, as we have merged generic mapper into master.
If we identify issues with this approach, we can add support for "qualifiers" as Romain mentioned...